Library rewrite of cookbook

cybermerc opened the pull request:
rewrote cookbook creating three libraries to make creation of master/slave configurations more uniform.

DO NOT MERGE TILL REVIEWED

Commit:
initial library build

Commit:
finish up version upgrade

Commit:
typo fix

Commit:
added circleci testing

brint commented on LICENSE:
```diff
@@ -0,0 +1,201 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright {yyyy} {name of copyright owner}
```
Update year, copyright owner

Commit:
typo

Commit:
update test suite

theborch commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@ pg-multi Cookbook CHANGELOG
 
+v1.0.0
+-------
+- Major rewrite, adding libraries as main use case vice recipes
```
> Major rewrite, adding libraries as main use case vice recipes
> ...use case, vice recipes.

theborch commented on Gemfile:
```diff
@@ -1,17 +1,15 @@
-  gem 'chef-zero'
 end
 
 group :kitchen_common do
```
> `group :kitchen_common do`

To keep inline with `kitchen_rackspace` this is probably more accurately named:
`group :kitchen_vagrant do`

cybermerc replied:
there is already a kitchen_rackspace and a kitchen_vagrant, the kitchen_common is for common gems needed by both

theborch commented on libraries/provider_pg_hba.rb:
```diff
@@ -0,0 +1,25 @@
+        new_resource.slave_ips.each do |slaveip|
+          node.default['postgresql']['pg_hba'] << {
+            comment: '# authorize slave server',
+            type: '#{new_resource.host_type}',
```
> `type: '#{new_resource.host_type}',`

String interpolation, use: `type: "#{new_resource.host_type}",`

theborch commented on libraries/provider_pg_hba.rb:
```diff
@@ -0,0 +1,25 @@
+            comment: '# authorize slave server',
+            type: '#{new_resource.host_type}',
+            db: 'replication',
+            user: '#{new_resource.repl_user}',
```
> `user: '#{new_resource.repl_user}',`

String interpolation, use: `user: "#{new_resource.repl_user}",`

theborch commented on libraries/provider_pg_hba.rb:
```diff
@@ -0,0 +1,25 @@
+            type: '#{new_resource.host_type}',
+            db: 'replication',
+            user: '#{new_resource.repl_user}',
+            addr: '#{slaveip}/32',
```
> `addr: '#{slaveip}/32',`

String interpolation, use: `addr: "#{slaveip}/32",`

theborch commented on libraries/provider_pg_repluser.rb:
```diff
@@ -0,0 +1,25 @@
+
+      action :create do
+        execute 'set-replication-user' do # ~FC009
+          role_exists = %(psql -c "SELECT rolname FROM pg_roles WHERE rolname='#{new_resource.repl_user}'" | grep #{new_resource.repl_user})
```
Have you considered doing this in ruby rather than shelling out, perhaps with the `pg` gem, along the lines of:

```
require 'pg'
psql = PG.connect( dbname: 'postgres' )
psql.exec("select rolname from pg_roles where rolname='#{new_resource.repl_user}'")
```

cybermerc replied:
if the pg gem would be useful anywhere else I would consider adding it, but for this its not worth the effort and added dependencies

theborch commented on metadata.rb:
```diff
@@ -1,11 +1,17 @@
+issues_url 'https://github.com/rackspace-cookbooks/pg-multi/issues'
 long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
-version '0.1.2'
+version '1.0.0' # bump this AFTER release, not in a PR or before
```
So isn't this being bumped in this PR?

cybermerc replied:
yes bumps in PR are required in major breaking change version bumps, that rule only applies on standard patch bumps

Commit:
sytax fixes

theborch commented on .kitchen.yml:
```diff
@@ -7,105 +7,22 @@ provisioner:
-    excludes: ['ubuntu-14.04', 'ubuntu-12.04']
+  - name: default
+    runlist:
+      - recipe[pg-multi::default]
```
> ```
> runlist:
>   - recipe[pg-multi::default]
> ```

Currently the tk suites are being generated with no run lists, all thanks to a missing `_` -- it's always the dang little things.
Correct version:

```
run_list:
  - recipe[pg-multi::default]
```

Commit:
fixed typo

martinb3 commented:
I know it isn't a perfect test, but could the test-kitchen stuff setup two instances and add replication between them? As far as I can tell, the test-kitchen default suite isn't even running any recipes except `pg-multi::default` which just calls `postgresql::server`. It seems like the bulk of the logic in this cookbook is in the other recipes, `pg_master` and `pg_slave`. I'd like to see the test-kitchen suite run at least one of them, if not both.

Neither the unit tests nor the serverspec tests actually run any of the providers we're writing in this cookbook, AFAICT.

martinb3 commented on README.md:
```diff
@@ -34,15 +49,18 @@ default of 8.4) you have to set this attribute as a node override:
 -----------
 
-`['pg-multi']['master']` : sets the IP address that defines the master node
+`['pg-multi']['master']` : sets the IP address that defines the master node (Required)
```
Are these correct? `attributes/default.rb` contains this (master vs. master_ip, slave vs. slave_ip):
`default['pg-multi']['master_ip'] = ''
default['pg-multi']['slave_ip'] = []`

cybermerc replied:
documentation updated, changed names forgot to update docs

martinb3 commented on README.md:
```diff
@@ -34,15 +49,18 @@ default of 8.4) you have to set this attribute as a node override:
 the replication user, by default it is set to 'repl'.
 
-`['pg-multi']['replication']['password'] ` : allows for the setting of a custom password for the replication user.
+`['pg-multi']['replication']['password'] ` : allows for the setting of a custom password for the replication user. (Required)
```
This does not appear to be required, according to `attributes/default.rb`:

```
default['pg-multi']['replication']['password'] = 'useagudpasswd'
```

cybermerc replied:
the "required" tags are for attributes that are required as part of the usage of the libraries. This is not well documented.

martinb3 replied:
The libraries don't use attributes, though, do they? They appear to have their own parameters with different names (e.g. `PgReplUser` and `PgSlave` use a parameter called `repl_pass`). I don't see `node['pg-multi']['replication']['password']` used anywhere in the libraries.

cybermerc replied:
True, they are not used directly but the attribute corresponds to a library resource, which is required. So yeah guess the 'required' tag is a bit misleading

martinb3 commented on recipes/pg_master.rb:
```diff
@@ -27,35 +27,25 @@
 when 'debian'
   node.set['pg-multi']['host'] = 'hostssl'
-else
+when 'redhat'
```
I believe the platform family for Red Hat is actually `rhel`.

cybermerc replied:
good catch, fixed

martinb3 commented on recipes/pg_master.rb:
```diff
@@ -27,35 +27,25 @@
 
-# set SSL usage based on OS support (debian yes, Redhat no)
+# since debian OS families can do ssl by default enable it, otherwise non-ssl connections
 case node['platform_family']
```
Since you are using chef-sugar in this recipe already, you might be interested in these: `debian?`, `rhel?` from https://github.com/sethvargo/chef-sugar/blob/master/lib/chef/sugar/platform_family.rb.

cybermerc replied:
done

martinb3 commented on recipes/pg_slave.rb:
```diff
@@ -28,15 +28,20 @@
 node.set['postgresql']['config']['wal_keep_segments'] = 8
 node.set['postgresql']['config']['hot_standby'] = 'on'
 
+case node['platform_family']
```
Same as above; chef-sugar has a shorter version.

cybermerc replied:
done

martinb3 commented on recipes/pg_slave.rb:
```diff
@@ -28,15 +28,20 @@
+case node['platform_family']
+when 'debian'
+  node.set['pg-multi']['host'] = 'hostssl'
+when 'redhat'
```
As in the other recipe, I think this should be `rhel`.

cybermerc replied:
good catch, fixed

martinb3 commented on libraries/provider_pg_hba.rb:
```diff
@@ -0,0 +1,25 @@
+
+      action :create do
+        new_resource.slave_ips.each do |slaveip|
+          node.default['postgresql']['pg_hba'] << {
```
I'm wondering if setting default attributes at converge time instead of compile time is going to surprise someone. Luckily, the upstream cookbook doesn't pass variables for the pg_hba.conf template -- it evaluates them at the time the template is rendered.

martinb3 commented on libraries/provider_pg_repluser.rb:
```diff
@@ -0,0 +1,25 @@
+      end
+
+      action :create do
+        execute 'set-replication-user' do # ~FC009
```
Do you need to wrap this in `converge_by` since you're saying `whyrun_supported? will return true`? From [this doc reference](https://docs.chef.io/lwrp.html#whyrun-supported):

> When why-run mode is supported by the a lightweight provider, the converge_by method is used to define the strings that are logged by the chef-client when it is run in why-run mode.

I can't tell if execute already supports 'why run' mode, but maybe it does. If it doesn't though, we'd need to wrap it.

> When a lightweight provider contains only platform resources, the converge_by method is not required because it is already built into all of the platform resources.

Maybe this means it isn't needed after all.

martinb3 commented on libraries/provider_pg_slave.rb:
```diff
@@ -0,0 +1,83 @@
+            rep_user:  new_resource.repl_user,
+            password:  new_resource.repl_pass
+          )
+          if new_resource.sensitive == 'true'
```
FWIW, I think you can do: `sensitive new_resource.sensitive` here instead of a conditional.

cybermerc replied:
fixed

martinb3 commented:
I tried to use this with chef search, which required a converge of the master to make the slave able to search for the master. Naturally, the slave converge failed because the master also needed to find a slave to allow it access in pg_hba.conf. This makes sense, since both nodes must know about each other to successfully converge. I could see other reasons why a converge might fail initially, but it led to a secondary issue:

If `bash[pull_master_databases]` fails, it still tries to run the other resources below that in the provider, such as `template "/var/lib/#{data_dir}/#{new_resource.pg_version}/main/recovery.conf"` (which then notifies the db service to restart).

Then it gets broken forever, because the next run of `postgresql::server_debian` tries to start the service on the next converge (it can't), so you never get another chance to run pg_basebackup (the inclusion of the upstream postgres recipes happens before the providers in this cookbook run again). So there's no way to repair the pair of servers using chef. You'd have to restore the slave again to a working state where the postgres service, in order to setup replication using chef again.

I'm concerned there might also be other situations in which a converge might fail, and then get into this same death spiral. Is there any way to be sure the resource below doesn't destructively break the service until we're sure the pull of data from the master succeeded?

```
        bash 'pull_master_databases' do
          user 'root'
          cwd '/tmp'
          code <<-EOH
          service postgresql stop
          sudo -u postgres rm -rf /var/lib/#{data_dir}/#{new_resource.pg_version}/main
          sudo -u postgres pg_basebackup -h #{new_resource.master_ip} -D /var/lib/#{data_dir}/#{new_resource.pg_version}/main -U repl -w --xlog-method=stream
          EOH
          not_if { ::File.exist?("/var/lib/#{data_dir}/#{new_resource.pg_version}/main/recovery.conf") }
        end
```

Those commands seem _very_ dangerous to do in an un-atomic way.

Commit:
additional cleanup

cybermerc commented:
cleaned up issues addressed. Also added a .kitchen.local.yml with configurations to be able to bring up master/slave systems via vagrant for further testing and evaluation.

martinb3 commented:
LGTM for merge. If we can get some atomicity in the bash script, :+1:  even better.
